### ISSUE TITLE
getInitialNotification always get the previous notification received

### ISSUE THREAD

**Issue by kenma9123** [evollu/react-native-fcm#305]
I've been pulling my hair out with this problem. The getInitialNotification event always gets the previous notification whenever you tap a notification from a tray. For both background and not running app. Everytime the app starts or opens, this gets emitted. I already clearing it using the ID but no joy.

What version of RN and react-native-fcm are you running? RN 0.40.0 / FCM 6.0.2
What device are you using? (e.g iOS9 emulator, Android 6 device)? Android emulator, Android 6 device
Is your app running in foreground, background or not running? background and not running

**Comment by evollu** [comment]
interesting, let me check

**Comment by kenma9123** [comment]
My temporary work around is to store the previous notification ID (coming from tray) to some store/state and then whenever the app starts and if the notification(the supposedly removed) is still there, I can check whether that notification was already delivered, and that seems to be working solution for me.

**Comment by evollu** [comment]
if you can debug android, you can add break point here and see why it happens. It will be very helpful. I can't find out by just looking at the code now.
https://github.com/evollu/react-native-fcm/blob/master/android/src/main/java/com/evollu/react/fcm/FIRMessagingModule.java#L59

**Comment by kenma9123** [comment]
what do you mean by adding breakpoint? To be honest I have a rusty java knowledge.

**Comment by evollu** [comment]
adding breakpoint in IDE, like android studio and run debug

**Comment by kenma9123** [comment]
I don't think I can help you with that. I only use atom as my editor no Android Studio or something.

**Comment by evollu** [comment]
no problem. I will take a look when I have time

**Comment by n8stowell82** [comment]
@evollu I am having this problem as well.  I have set a breakpoint where you have indicated, and it is getting hit as soon as the application is launched.  I see a promise passed in as a param, but I don't know what I am looking for to help you debug.  Please let me know what I can be looking for.  

Is there not a way to clear the initial notification once it has been read?  I am calling `FCM.removeAllDeliveredNotifications();` now as soon as I have read the initialNotification.

BTW I am running RN 0.35 and v2.5.6 of this library

**Comment by evollu** [comment]
Ok I will check

**Comment by evollu** [comment]
Initial notification is the notification that launches the app. It will remain unchanged through the whole app lifetime.
When app is in background and user click the banner to resume the app, there will be a notification event triggered with special flag "open_from_tray".
When app is not running and user click the banner to launch the app, you should get notification data in initial notification

I've checked behavior and seems both background and killed app work as expected.

**Comment by n8stowell82** [comment]
Ok, I may need to test this further.  I recently upgraded to RN 41 and the latest of this repo.  What could possibly be happening is that I am just refreshing the JS package each time (in dev mode) and that is not actually flushing the initial notification from the native side.  What I was seeing was that once I received a notification, initial notification was firing every time despite having "read" that message.  My memory is a bit fuzzy right now, but I am wondering if it is just when the js refreshes is not clearing that flag.  I will try and test this tomorrow (Monday) and report back. thanks for your time

**Comment by Msspl-KrishnenduBhattacharyya** [comment]
Yes,
I am facing same problem. After using FCM.getInitialNotification(), I used FCM.removeAllDeliveredNotifications() but the same problem I am facing.

**Comment by krishbhattacharyya** [comment]
Every moment, When I open the closed app it triggers  FCM.getInitialNotification() but it should not trigger.  

[ As per documentation; When app is not running and user clicks notification, notification data will be passed into FCM.getInitialNotification event  ]

But for the quick solution I did: 
FCM.getInitialNotification().then((notif)=> {
        // is_push is a flag passing in data 
        if(notif.is_push){
            // Your code
            // like I used this.props.navigator.push({});
        }
}); 

May be it can help temporary. FCM.removeAllDeliveredNotifications() should work.

**Comment by evollu** [comment]
@krishbhattacharyya `FCM.getInitialNotification()` is not a event listener, rather a async function call. If you trigger it every time resume the app, it means the function gets called. Move it to root component

**Comment by krishbhattacharyya** [comment]
Thank you @evollu 

I think the documentation should be updated, you have described beautifully. 

Specially this portion:
[ When app is not running and user clicks notification, notification data will be passed into FCM.getInitialNotification event ]

**Comment by evollu** [comment]
will update

**Comment by mradziwon** [comment]
Is there any way to clear this initial notification besides killing the app?

**Comment by andresmeidla** [comment]
@mradziwon calling notification.finish(); will clear the notification.

**Comment by andresmeidla** [comment]
actually scratch that, finish() is not available for the initial notification :/

**Comment by diegorodriguesvieira** [comment]
@andresmeidla do you have any workaround to finish initial notification?

**Comment by andresmeidla** [comment]
@diegorodriguesvieira got a workaroound for android only atm. in my case, im launching the app from a data notification myself using a launch intent. I just added the FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS flag to the launch intent, so it wouldn't be possible to launch the the same intent from the recent apps list after exiting from the app.

Haven't tried iOS yet, though, so don't know how to work around that yet.

**Comment by nkov** [comment]
@radev @andresmeidla I think what @evollu was saying in his previous comment was if you move `FCM.getInitialNotification()` from `componentDidMount()` to outside of the component entirely (I have mine above the component class definition for my main component), then the event fires when I open it from a notification, and if I minimize and open it again, it's either not displaying the initial notification at all or showing that it was started with the main intent, and I don't need to catch it.

For now also tried only in Android.

**Comment by beeglebug** [comment]
This seems related to #339, a lot of people are seeing weird calls to getInitialNotification, even when it is only called once on app load, and when there are no actual notifications in the tray which need delivering to the app

**Comment by evollu** [comment]
@nkov might be right. `FCM.getInitialNotificaiton()` can be placed outside of component lifecycle. May be in redux init?

**Comment by jqn** [comment]
+1 I running into the same issue, is this a misunderstanding about how this library works or is it a bug? anyone found a good solution yet?

**Comment by jqn** [comment]
I figured out most of the issue here is due to a misunderstanding on how this library works. Moving FCM.getInitialNotificaiton() and  FCM.on(FCMEvent.Notification, async notif => {
  console.log('notification payload', notif);
}); out of the component lifecycle works the way it is expected with no problems. At least for me.

Example: 
```
import {
  AppState,
  Image,
  TouchableOpacity,
  Text,
  View
} from 'react-native';

import FCM, {
  FCMEvent,
  RemoteNotificationResult,
  WillPresentNotificationResult,
  NotificationType
} from 'react-native-fcm';

FCM.getInitialNotification().then(notification => {
  console.log('initial notification', notification);
});

FCM.on(FCMEvent.Notification, async notif => {
  console.log('notification payload', notif);
});

class PushNotificationManager extends Component {
  ....
}

export default PushNotificationManager;
```

**Comment by beeglebug** [comment]
I think that's a separate issue.

Even outside of component lifecycle methods it still triggers every time
the app boots, so everyone who uses the library needs to write similar
boilerplate to filter the spurious calls which don't actually contain a
real notification.

**Comment by jqn** [comment]
correct, after the app is killed getInitialNotification returns the following:

![screen shot 2017-11-30 at 4 06 11 pm](https://user-images.githubusercontent.com/9710893/33459736-67a5d57a-d5e8-11e7-9159-f869fd02eb7d.png)

this still needs to be handled with a special filter but the previously received notification is not longer there making that process easier.

**Comment by Ashwin-Mothilal** [comment]
But in the example sample-fcm-client the initial notif is stored in a state. So the thing is where should we use it? In the parent Navigator or outside of component or using it in a state?

**Comment by evollu** [comment]
@Ashwin-Mothilal it is up to you. it doesn't have to be in a state if you consume it right away

**Comment by gianpaj** [comment]
While I'm not using this library (I'm using `react-native-firebase`) I was getting the same issue as the title. 

As a workaround, I'm storing the `notificationId` after navigating to a screen:

Note: You might need to get the notification id in a different way for `react-native-fcm`

```js
import { AsyncStorage } from 'react-native';

...
// react-native-firebase (not react-native-fcm)
firebase
  .notifications()
  .getInitialNotification()
  .then(async (notificationOpen) => {
    console.log('getInitialNotification');

    // App was opened by a notification
    if (notificationOpen) {
      // Get information about the notification that was opened
      const notification = notificationOpen.notification;
      const lastNotification = await AsyncStorage.getItem(
        'lastNotification'
      );
      if (lastNotification !== notification.notificationId) {
        navigate(notification);
        await AsyncStorage.setItem(
          'lastNotification',
          notification.notificationId
        );
      }
    }
  })
```

Hope this is helpful for others

**Comment by juanfer0002** [comment]
Hi, any update on this issue?

**Comment by RishoniFG** [comment]
anyone found a good solution yet?

**Comment by krutika211** [comment]
Is there any update for this issue

**Comment by kennethpdev** [comment]
There are few workarounds, just explore. My workaround is still kicking. Store the notification somewhere(via id) then when the app boots and the `getInitialNotification` triggered. Compare that to the one you just stored. If its the same then just ignore it.

**Comment by namnm** [comment]
Recently I face this issue.
I have a fix here if anyone interest:
https://github.com/evollu/react-native-fcm/compare/master...brekekesoftware:master

**Comment by jamalud-lab** [comment]
+1 same issue
